common/src/main/java/io/netty/util/internal/RefCnt.java

```diff
@@ -0,0 +1,476 @@
+/*
+ * Copyright 2025 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal;
+
+import io.netty.util.IllegalReferenceCountException;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import static io.netty.util.internal.ObjectUtil.checkPositive;
+
+/**
+ * Monomorphic reference counter implementation that always use the most efficient available atomic updater.
+ * This implementation is easier for the JIT compiler to optimize,
+ * compared to when {@link ReferenceCountUpdater} is used.
+ */
+@SuppressWarnings("deprecation")
+public class RefCnt {
```

Comment:
hmmm after thinking a bit more about it wouldn't that basically break again on older android versions:

https://github.com/netty/netty/issues/15654

Reply:
I think it's fine because all VarHandle uses happen in VarHandleRefCnt which is a separate class that is never loaded when VarHandle is not supported